### PR TITLE
Create eyeball-im-util with FilteredVectorSubscriber API

### PR DESCRIPTION
**Pull request by jplatte**

### CODE REVIEW

File: eyeball-im-util/src/vector.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            new_value_matches.then(|| {
+                self.filtered_indices.insert(index, original_idx);
+                VectorDiff::Insert { index, value }
+            })
+        }
+    }
+
+    fn handle_remove(mut self: Pin<&mut Self>, index: usize) -> Option<VectorDiff<T>> {
+        let original_idx = index;
+        self.original_len -= 1;
```
Comment (bnjbvr) [inline eyeball-im-util/src/vector.rs:169]: Ditto for check here.

**Review by bnjbvr — COMMENTED** [review]
Looks interesting! I've found a few places where there could be invalid subtracts (unless I'm missing something), can you add tests for those please?

Also, I think there might be a way to make all this code more robust, by testing invariants. Say something like `fn assert_invariants`, that would be either debug only (or guarded behind a compile-time cargo feature, enabled in tests only), that would check the following at the end of each `handle_` function call:

- indices in `filtered_indices` are unique (don't appear twice or more)
- indices in `filtered_indices` refer to values that are in the original array, *and* those values pass the filter check

What do you think? This might mean storing more of the original array too, but at least for testing purposes, something akin to that would be useful and increase the confidence in this new code IMO.

File: eyeball-im-util/src/vector.rs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        let original_idx = self.original_len;
+        self.original_len += 1;
+        (self.filter)(&value).then(|| {
+            self.filtered_indices.push_back(original_idx);
+            VectorDiff::PushBack { value }
+        })
+    }
+
+    fn handle_pop_front(mut self: Pin<&mut Self>) -> Option<VectorDiff<T>> {
+        self.original_len -= 1;
```
[the remark this answers is not in the document]
Comment (jplatte) [reply eyeball-im-util/src/vector.rs:112]: If we received a `VectorDiff::PopFront` with the original len being 0, that would be a bug in `eyeball-im` and we couldn't really do anything more useful than panicking.

File: eyeball-im-util/src/vector.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        let result = self.filtered_indices.front().map_or(false, |&idx| idx == 0).then(|| {
+            assert!(self.filtered_indices.pop_front().is_some());
+            VectorDiff::PopFront
+        });
+        for idx in &mut self.filtered_indices {
+            *idx -= 1;
+        }
```
[the remark this answers is not in the document]
Comment (jplatte) [reply eyeball-im-util/src/vector.rs:119]: What else could a loop over an empty list do, than run the body zero times?

[the remark this answers is not in the document]
Comment (bnjbvr) [reply eyeball-im-util/src/vector.rs:119]: I had one eyebrow naturally raise ("what if idx is zero") but that's impossible, fair enough.

**Comment by jplatte** [comment]
I think testing invariants would be nice, but I don't think doing it as part of the regular non-`cfg(test)` is a good idea, since app debug builds by default also compile all dependencies in debug mode.

A good way to do invariant testing would be adding another test suite IMO (as unit tests that can access internal stat), using the [`proptest`](https://docs.rs/proptest/latest/proptest/) crate to ensure the internal state is kept consistent in ~all cases. There's also the option of trying [Kani](https://github.com/model-checking/kani#readme), but it's probably a bit harder to do and possibly an overkill.

I'd prefer not blocking this PR on extra testing though.. We don't intend to use it in places where it would be a huge issue for some bugs to be present, right? And if there are bugs, they'll likely be pretty obvious (i.e. lead to a panic).

**Review by Hywan — APPROVED** [review]
I probably like it too much, thank you!

**Pull request merged**